### PR TITLE
[rdy] Fix cheat system

### LINKED ISSUE #1718: Recode Cheats out of UI
As talked about with @Alberth289346 we want to move Cheats so it is not relying on UI.

I need to look into it further before doing so but don't have time at the moment. This is why PR #1719  is in to stop trunk breaking for the moment.

### PR DESCRIPTION
**Pull request by lewri**
*Fixes #1718*

**Describe what the proposed change does**
- Moves cheat instantiation into player_hospital.
- Removes dependency on ui (in all but one case).

NB: An afterload in hospital.lua is deleted. But I'm not sure if a new afterLoad is needed to clear hosp_cheats if it was already created (i.e. would it interfere with the hosp_cheats in player_hospital.lua?)

### CODE REVIEW

**Comment by TheCycoONE** [comment]
Your afterLoad looks fine to me.

**Comment by lewri** [comment]
This is ready but I would prefer is someone could confirm independently this is working first.

**Comment by tobylane** [comment]
All good for me.

**Comment by lewri** [comment]
Thanks, good to go then whenever you get time cyco :)

File: CorsixTH/Lua/cheats.lua [hunk 1]
```diff
@@ -34,24 +34,33 @@ function Cheats:Cheats(hospital)
...
-    {name = "vip",            func = --[[persistable:cheatVip]] function() self:cheatVip() end},
-    {name = "earthquake",     func = --[[persistable:cheatEarthquake]] function() self:cheatEarthquake() end},
-    {name = "create_patient", func = --[[persistable:cheatPatient]] function() self:cheatPatient() end},
-    {name = "end_month",      func = --[[persistable:cheatMonth]] function() self:cheatMonth() end},
-    {name = "end_year",       func = --[[persistable:cheatYear]] function() self:cheatYear() end},
-    {name = "lose_level",     func = --[[persistable:cheatLose]] function() self:cheatLose() end},
-    {name = "win_level",      func = --[[persistable:cheatWin]] function() self:cheatWin() end},
-    {name = "increase_prices", func = --[[persistable:cheatIncreasePrices]] function() self:cheatIncreasePrices() end},
-    {name = "decrease_prices", func = --[[persistable:cheatDecreasePrices]] function() self:cheatDecreasePrices() end},
+   self.cheat_list = {
```
Comment (Alberth289346) [inline CorsixTH/Lua/cheats.lua:37]:
```suggestion
  self.cheat_list = {
```

File: CorsixTH/Lua/cheats.lua [hunk 2]
```diff
@@ -34,24 +34,33 @@ function Cheats:Cheats(hospital)
...
+    {name = "toggle_infected", func = self.cheatToggleInfected},
+    {name = "vip",            func = self.cheatVip},
+    {name = "earthquake",     func = self.cheatEarthquake},
+    {name = "create_patient", func = self.cheatPatient},
+    {name = "end_month",      func = self.cheatMonth},
+    {name = "end_year",       func = self.cheatYear},
+    {name = "lose_level",     func = self.cheatLose},
+    {name = "win_level",      func = self.cheatWin},
+    {name = "increase_prices", func = self.cheatIncreasePrices},
+    {name = "decrease_prices", func = self.cheatDecreasePrices},
```
Comment (Alberth289346) [inline CorsixTH/Lua/cheats.lua]:
```suggestion
    {name = "money",           func = self.cheatMoney},
    {name = "all_research",    func = self.cheatResearch},
    {name = "emergency",       func = self.cheatEmergency},
    {name = "epidemic",        func = self.cheatEpidemic},
    {name = "toggle_infected", func = self.cheatToggleInfected},
    {name = "vip",             func = self.cheatVip},
    {name = "earthquake",      func = self.cheatEarthquake},
    {name = "create_patient",  func = self.cheatPatient},
    {name = "end_month",       func = self.cheatMonth},
    {name = "end_year",        func = self.cheatYear},
    {name = "lose_level",      func = self.cheatLose},
    {name = "win_level",       func = self.cheatWin},
    {name = "increase_prices", func = self.cheatIncreasePrices},
    {name = "decrease_prices", func = self.cheatDecreasePrices},
```

File: CorsixTH/Lua/cheats.lua [hunk 3]
```diff
@@ -34,24 +34,33 @@ function Cheats:Cheats(hospital)
...
   }
 end
 
+--! Performs a cheat from the cheat_list
+--!param num The cheat from the cheat_list called
```
Comment (Alberth289346) [inline CorsixTH/Lua/cheats.lua:56]:
```suggestion
--!param num (integer) The cheat from the cheat_list called
```

**Review by Alberth289346 — COMMENTED** [review]
Some minor indentation and documentation fixes

File: CorsixTH/Lua/cheats.lua [hunk 4]
```diff
@@ -34,24 +34,33 @@ function Cheats:Cheats(hospital)
...
+--!return true if cheat was successful, false otherwise
+function Cheats:performCheat(num)
+  local cheat_success = self.cheat_list[num].func(self) ~= false
+  return cheat_success and self.cheat_list[num].name ~= "lose_level"
```
Comment (TheCycoONE) [inline CorsixTH/Lua/cheats.lua:60]: Future - we shouldn't have to special case loseLevel.

Comment (lewri) [reply CorsixTH/Lua/cheats.lua:60]: Did you want this added as a comment or general note for the PR?

File: CorsixTH/Lua/cheats.lua [hunk 5]
```diff
@@ -73,24 +83,24 @@ function Cheats:cheatResearch()
 end
 
 function Cheats:cheatEmergency()
-  if not self.ui.hospital:createEmergency() then
-    self.ui:addWindow(UIInformation(self.ui, {_S.misc.no_heliport}))
+  if not self.hospital:createEmergency() then
```
Comment (mugmuggy) [inline CorsixTH/Lua/cheats.lua:86]: This isn't something you've introduced, but you can effectively neutralise active epidemics, by starting an emergency here, as the UIWatch check is bypassed at this point. However patients will still be infected/contagious (can't vaccinate them as you fixed this previosuly) well after the natural end, as the 'interrupted' epidemic never finishes.

Comment (TheCycoONE) [reply CorsixTH/Lua/cheats.lua:86]: Thanks @mugmuggy , split that out as a separate issue to address later.

**Pull request merged**